include/tvm/relay/pass.h

```diff
@@ -2,18 +2,202 @@
+ *
+ * The responsibilities of a pass manager usually involves:
+ *  - Organizing the execution order of optimization passes though not
+ * necessarily in the optimal sequence.
+ *  - Collecting required analysis information and keep them up-to-date.
+ *  - Reducing the effort required to implement new passes for compiler
+ * developers, etc.
+ *
+ * TODO(jroesch, zhiics): We are currently using a very simple design for the
+ * pass manager, i.e. it executes a specific pass or sequence of passes.
+ *
+ * In the future we need to describe constraints between passes. For example,
+ * we may want to preserve dependencies between different passes and validate
+ * them on the completion of a certain pass.
+ *
+ * We also need to store side information and import the error reporting system.
  */
 #ifndef TVM_RELAY_PASS_H_
 #define TVM_RELAY_PASS_H_
 
+#include <tvm/attrs.h>
+#include <tvm/ir.h>
+#include <tvm/packed_func_ext.h>
+#include <tvm/relay/error.h>
 #include <tvm/relay/expr.h>
 #include <tvm/relay/module.h>
 #include <tvm/relay/op_attr_types.h>
+#include <tvm/relay/type.h>
+
 #include <string>
+#include <vector>
 
 namespace tvm {
 namespace relay {
 
+namespace pass {
+
+// Forward declaration
+class ModulePass;
+class FunctionPass;
+class SequentialPass;
+
+// Define pass context.
+class PassContext;
+
+/*!
+ * \brief PassContextNode contains the information that a pass can rely on, such as
+ * analysis results.
+ */
+class PassContextNode : public RelayNode {
+ public:
+  /*!
+   * \brief The error reporter used to notify users why an optimization fails.
+   */
+  ErrorReporter err_reporter_;
+
+  PassContextNode() = default;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+  }
+
+  TVM_DLL static PassContext make();
+
+  static constexpr const char* _type_key = "relay.PassContext";
+  TVM_DECLARE_NODE_TYPE_INFO(PassContextNode, RelayNode);
+};
+
+class PassContext : public NodeRef {
+ public:
+  PassContext() = default;
+  explicit PassContext(NodePtr<tvm::Node> p) : NodeRef(p) {}
+
+  const PassContextNode* operator->() const {
+    return static_cast<PassContextNode*>(this->node_.get());
+  }
+
+  using ContainerType = PassContextNode;
+};
+
+// We use currying here. It runs on a Relay node type NodeT and yields a new
+// node with the same type. The Relay module is captured for optimizations as
+// most of the current Relay optimizations are module to module. Currying
+// sketches the optimization, i.e. how we want to mutate an AST, and it is
+// passed as packed functions that will be invoked when called by `run`.
+//
+// For example, PassFunc<Function> indicates we perform a Function to Function
+// transformation on the given Module.
+template <typename NodeT,
+          typename = std::enable_if<(std::is_same<NodeT, Module>::value ||
+                                     std::is_same<NodeT, Function>::value)>>
+using PassFunc =
+    runtime::TypedPackedFunc<runtime::TypedPackedFunc<NodeT(NodeT)>(
```

Comment:
double check the signature. The signature may no longer be valid. We should avoid capturing the relay module. Instead we might want to simply do ```TypedPackedFunc<NodeT(NodeT, PassContext)>```. because it is simple enough, we can remove PassFunc's definition and inline

Reply:
yeah, I think we may still need it. Otherwise, we will probably need to move `Module` to `PassContext` for capturing which would be the same like the old `PassState`.

@jroesch thoughts?

Reply:
We can do (NodeT, PassContext)->NodeT, so you do not capture the module, the outer pass will call this function to transform the corresponding module.

Reply:
@tqchen hmm, I forgot the module is already passed in the functor. I think the updated version is what you meant, right? But in this case, I think we can probably also ignore PassContext as it is already a member of a `Pass`, is it correct?